### PR TITLE
Deeplabcut - update pytorch params

### PR DESCRIPTION
**Pull request by kapilseh**
- Modified a few PyTorch params in the PyTorch config file
- Removes the following error msg from log   
image

### CODE REVIEW

File: DeepLabCut/deeplabcut_main.py [hunk 1]
```diff
@@ -82,6 +83,25 @@ def update_config_file(config_file_path, bodypart_names):
     with open(config_file_path, 'w') as file:
         yaml.safe_dump(data, file, default_flow_style=False)
 
+def update_Pytorch_config_file(config_file_path):
+    """
+    Update label names in the pytorch config file
+    """
+    root_path   = os.path.dirname(config_file_path)
+    target_file = 'pytorch_config.yaml'
+    for dir_path, dirnames, filenames in os.walk(root_path):
```
Comment (4D42) [inline DeepLabCut/deeplabcut_main.py:92]: Shouldn't the file ` 'pytorch_config.yaml' ` always be at the same place?

File: DeepLabCut/deeplabcut_main.py [hunk 2]
```diff
@@ -82,6 +83,25 @@ def update_config_file(config_file_path, bodypart_names):
     with open(config_file_path, 'w') as file:
         yaml.safe_dump(data, file, default_flow_style=False)
 
+def update_Pytorch_config_file(config_file_path):
```
Comment (doriclenses) [inline DeepLabCut/deeplabcut_main.py:86]: any reason the P is capital?

**Review by doriclenses — CHANGES_REQUESTED** [review]
Please resolve the comment of Maxime

File: DeepLabCut/deeplabcut_main.py [hunk 3]
```diff
@@ -34,7 +34,8 @@ def main(deeplabcut_params: dlc_params.DeepLabCutParameters):
     # Create project and train network
     video_path, config_file_path = create_project(filepath, datapath, project_folder, bodypart_names, extracted_frames, deeplabcut_params.params)
     deeplabcut.create_training_dataset(config_file_path)
-    deeplabcut.train_network(config_file_path)
+    update_pytorch_config_file(config_file_path)
+    deeplabcut.train_network(config_file_path, batch_size=8)
```
Comment (doriclenses) [inline DeepLabCut/deeplabcut_main.py:38]: To be sure, is this `batch_size` related to the training or image augmentation?
image

Comment (kapilseh) [reply DeepLabCut/deeplabcut_main.py:38]: its for Pytorch train setting.
image

File: DeepLabCut/deeplabcut_main.py [hunk 4]
```diff
@@ -82,6 +83,25 @@ def update_config_file(config_file_path, bodypart_names):
...
+    for dir_path, dirnames, filenames in os.walk(root_path): 
+        if target_file in filenames:
+            print(dir_path)
+            pytorch_config_file_path = os.path.join(dir_path, target_file)
+
+    with open(pytorch_config_file_path, 'r') as file:
+        data = yaml.safe_load(file)
+  
+    data['model']['backbone']['freeze_bn_stats']    = False
+    data['train_settings']['dataloader_pin_memory'] = True
```
[the remark this answers is not in the document]
Comment (kapilseh) [reply DeepLabCut/deeplabcut_main.py:101]: the freeze_Stat change helps remove the warning -
image

and  data['train_settings']['dataloader_pin_memory'] = True  is recommended when using nvidia GPU
image

File: DeepLabCut/deeplabcut_main.py (hunk 3 above)
Comment (doriclenses) [reply DeepLabCut/deeplabcut_main.py:38]: image
image
image
Still not sure...

File: DeepLabCut/deeplabcut_main.py [hunk 5]
```diff
@@ -33,14 +33,21 @@ def main(deeplabcut_params: dlc_params.DeepLabCutParameters):
...
-    deeplabcut.train_network(config_file_path)
-    deeplabcut.evaluate_network(config_file_path)
- 
+    training_dataset_info = deeplabcut.create_training_dataset(config_file_path)
+    
+    shuffle_nums = len(training_dataset_info)
+    shuffle_index_start = training_dataset_info[0][1]
+    shuffles = [shuffle_index_start + i for i in range(shuffle_nums)]
+
+    for shuffle_index in shuffles:
```
Comment (4D42) [inline DeepLabCut/deeplabcut_main.py:42]: I think you can do `shuffles = shuffle_index_start + np.arange(shuffle_nums)` or maybe something like `shuffles = list(range(shuffle_index_start, shuffle_index_start + shuffle_nums))`

**Review by 4D42 — APPROVED** [review]

File: DeepLabCut/deeplabcut_main.py [hunk 6]
```diff
@@ -33,14 +33,21 @@ def main(deeplabcut_params: dlc_params.DeepLabCutParameters):
...
+    shuffles = list(range(shuffle_index_start, shuffle_index_start + shuffle_nums))
+
+    for shuffle_index in shuffles:
+        update_pytorch_config_file(config_file_path, shuffle_index) 
+        deeplabcut.train_network(config_file_path, batch_size=8, shuffle = shuffle_index)
+    
+    deeplabcut.evaluate_network(config_file_path, Shuffles = shuffles)
     # Analyze video and save the result
-    deeplabcut.analyze_videos(config_file_path, [video_path], destfolder = os.path.dirname(config_file_path))
+    deeplabcut.analyze_videos(config_file_path, [video_path], destfolder = os.path.dirname(config_file_path), shuffle = shuffle_index_start)
```
Comment (doriclenses) [inline DeepLabCut/deeplabcut_main.py:48]: If we train network with the `shuffle = 2`, this would still analyze the video with the `shuffle = 1`, no?

File: DeepLabCut/deeplabcut_main.py [hunk 7]
```diff
@@ -33,14 +33,21 @@ def main(deeplabcut_params: dlc_params.DeepLabCutParameters):
...
     video_path, config_file_path = create_project(filepath, datapath, project_folder, bodypart_names, extracted_frames, deeplabcut_params.params)
-    deeplabcut.create_training_dataset(config_file_path)
-    deeplabcut.train_network(config_file_path)
-    deeplabcut.evaluate_network(config_file_path)
- 
+    training_dataset_info = deeplabcut.create_training_dataset(config_file_path)
+    
+    shuffle_nums = len(training_dataset_info)
+    shuffle_index_start = training_dataset_info[0][1]
+    shuffles = list(range(shuffle_index_start, shuffle_index_start + shuffle_nums))
```
[the remark this answers is not in the document]
Comment (kapilseh) [reply DeepLabCut/deeplabcut_main.py:40]: create_training_dataset returns ->  training_dataset_info  which is ---> 
 list(tuple) or None
        If training dataset was successfully created, a list of tuples is returned.
        The first two elements in each tuple represent the training fraction and the
        shuffle value. The last two elements in each tuple are arrays of integers
        representing the training and test indices.

something like this 
training_dataset_info = [(0.95, 1, (array([2, 1]), array([0])))]

File: DeepLabCut/deeplabcut_main.py [hunk 8]
```diff
@@ -33,14 +33,17 @@ def main(deeplabcut_params: dlc_params.DeepLabCutParameters):
...
-    deeplabcut.create_training_dataset(config_file_path)
-    deeplabcut.train_network(config_file_path)
-    deeplabcut.evaluate_network(config_file_path)
- 
+    training_dataset_info = deeplabcut.create_training_dataset(config_file_path)
+    
+    shuffle_index = training_dataset_info[0][1]
+    update_pytorch_config_file(config_file_path, shuffle_index) 
+    deeplabcut.train_network(config_file_path, batch_size=8, shuffle = shuffle_index)
+    deeplabcut.evaluate_network(config_file_path, Shuffles = [shuffle_index])
```
Comment (doriclenses) [inline DeepLabCut/deeplabcut_main.py:41]: I don't think we need to set the argument `Shuffles`

Comment (kapilseh) [reply DeepLabCut/deeplabcut_main.py:41]: if we don't put it it will use the default 1
def evaluate_network(
    config,
    Shuffles: Iterable[int] = (1,),

Shuffles: list, optional, default=[1]
        List of integers specifying the shuffle indices of the training dataset.

File: DeepLabCut/deeplabcut_main.py [hunk 9]
```diff
@@ -132,7 +161,7 @@ def create_labeled_data(config_file_path, extracted_frames, bodypart_names, expe
     cap.release() 
     cv2.destroyAllWindows()
 
-def save_coords_to_doric(filepath, datapath, output_path, params, group_names):
+def save_coords_to_doric(filepath, datapath, output_path, params, config_file_path, shuffle_index, group_names):
```
Comment (doriclenses) [inline DeepLabCut/deeplabcut_main.py:164]: If you pass `config_file_path`, do we still need the `output_path`?

File: DeepLabCut/deeplabcut_main.py [hunk 10]
```diff
@@ -33,14 +33,17 @@ def main(deeplabcut_params: dlc_params.DeepLabCutParameters):
 
     # Create project and train network
     video_path, config_file_path = create_project(filepath, datapath, project_folder, bodypart_names, extracted_frames, deeplabcut_params.params)
-    deeplabcut.create_training_dataset(config_file_path)
-    deeplabcut.train_network(config_file_path)
-    deeplabcut.evaluate_network(config_file_path)
- 
+    training_dataset_info = deeplabcut.create_training_dataset(config_file_path)
+    
+    shuffle_index = training_dataset_info[0][1]
```
Comment (doriclenses) [inline DeepLabCut/deeplabcut_main.py:38]: Could you just name `shuffle` as it's named in deeplabcut?

File: DeepLabCut/deeplabcut_main.py [hunk 11]
```diff
@@ -160,10 +189,33 @@ def save_coords_to_doric(filepath, datapath, output_path, params, group_names):
...
     utils.save_attributes(utils.merge_params(params_current = params), file_, operation_path)
 
-    # Save coordinates for each body part
-    h5_file1 = glob.glob(os.path.join(output_path, '*.h5'))[0]
-    df_coords = pd.read_hdf(h5_file1)
+    root_path   = os.path.dirname(config_file_path)
+    target_file = 'pytorch_config.yaml'
+
+    # get info from configFile
+    with open(config_file_path, 'r') as file:
```
Comment (doriclenses) [inline DeepLabCut/deeplabcut_main.py:196]: Lines 96-103 and 196-203 are exactly the same, so they should be changed to a function in the next pull request. I'll approve this one.

**Review by doriclenses — APPROVED** [review]
